### PR TITLE
Made runtime=compiled the default

### PR DESCRIPTION
**Pull request by pontusmelke**
Fixed some bugs in compiled plans showing up when running all tests
- Explain in compiled methods
- Proper error handling of user specified exceptions and `KernelExceptions`
- Fixed statement handling on when closing iterators
- Rewrote `dumpToString` such that it uses `accept` instead of `toList`  
- Forced some tests to use interpreted runtime which uses `PROFILE`

### CODE REVIEW

File: community/cypher/cypher/src/test/scala/org/neo4j/cypher/NewPlannerTestSupport.scala [hunk 1]
```diff
@@ -211,9 +220,11 @@ trait NewPlannerTestSupport extends CypherTestSupport {
...
     //if action fails we must wait to throw until after test has run
     val result = Try(action) match {
-      case f@Failure(ex: CantHandleQueryException) =>
-        testPlanner(newPlannerMonitor.trace)
-        testRuntime(newRuntimeMonitor.trace)
+      case Failure(ex: CantHandleQueryException) =>
+        // this should never happen, but make the thing to compile:
+        throw ex
+
+      case Failure(ex: CantCompileQueryException) =>
```
Comment (davidegrohmann) [inline community/cypher/cypher/src/test/scala/org/neo4j/cypher/NewPlannerTestSupport.scala]: this doesn't look right

File: community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/birk/CompiledMathHelper.java [hunk 2]
```diff
@@ -51,12 +52,39 @@ public static Object add( Object lhs, Object rhs )
...
+        }
+        else if (lhs instanceof List<?>)
+        {
+            List<?> lhsList = (List<?>) lhs;
+            List<Object> result = new ArrayList<>( lhsList.size() + 1 );
+            result.addAll( lhsList );
+            result.add(rhs);
+            return result;
+        }
+        else if (rhs instanceof List<?>)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/birk/CompiledMathHelper.java]: Code style for ´if´ should have padding space inside parenthesis.

File: community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/birk/CompiledMathHelper.java [hunk 3]
```diff
@@ -51,12 +52,39 @@ public static Object add( Object lhs, Object rhs )
...
+        {
+            List<?> rhsList = (List<?>) rhs;
+            List<Object> result = new ArrayList<>( rhsList.size() + 1 );
+            result.add(lhs);
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/birk/CompiledMathHelper.java]: Code style for `if` and method invocation should have padding space inside parenthesis (five occurrences above this comment). Not sure about casts; are they with or without padding?

File: community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/birk/Javac.java [hunk 4]
```diff
@@ -99,13 +98,13 @@ public CompilationError( String message )
         return clazz;
     }
 
-    public static InternalExecutionResult newInstance( Class<InternalExecutionResult> clazz, CompletionListener completion, Statement statement,
+    public static InternalExecutionResult newInstance( Class<InternalExecutionResult> clazz, TaskCloser closer, Statement statement,
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/java/org/neo4j/cypher/internal/compiler/v2_3/birk/Javac.java:101]: This is going to conflict with the profiling PR. I say we merge this one first, and I can rebase the other one.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/codegen/LogicalPlanConverter.scala [hunk 5]
```diff
@@ -182,7 +182,9 @@ object LogicalPlanConverter {
       (methodHandle, ProjectProperties(projectionInstructions, action))
     }
 
-    private def createProjectionInstruction(expression: Expression, context: CodeGenContext): ProjectionInstruction =
+    private def createProjectionInstruction(expression: Expression, context: CodeGenContext): ProjectionInstruction = {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/codegen/LogicalPlanConverter.scala:185]: Return type annotation should not be needed here (given by name anyway).

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala [hunk 6]
```diff
@@ -28,16 +30,20 @@ trait Instruction {
...
 
-  // Initialises necessary data-structures. Is inserted at the top of the generated method
+  //Initialises necessary data-structures. Is inserted at the top of the generated method
   def generateInit(): String
 
   def methods: Seq[Method] = {
     (allLeafs :+ this).flatMap(_._method)
   }
 
+  def exceptions: Set[ExceptionCodeGen] = allLeafs.flatMap(_._exceptions()).toSet
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala:40]: Make final

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala [hunk 7]
```diff
@@ -58,5 +64,7 @@ object Instruction {
     override def fields(): String = ""
 
     override def generateInit(): String = ""
+
+    override def exceptions: Set[ExceptionCodeGen] = Set.empty
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala:68]: It should override the underscore version, shouldn't it? I understand that `f` is the aggregating function and `_f` is the "local" function, is this correctly interpreted?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala [hunk 8]
```diff
@@ -28,16 +30,20 @@ trait Instruction {
...
   def methods: Seq[Method] = {
     (allLeafs :+ this).flatMap(_._method)
   }
 
+  def exceptions: Set[ExceptionCodeGen] = allLeafs.flatMap(_._exceptions()).toSet
+
   // Generates import list for class
   protected def _importedClasses(): Set[String] = Set.empty
 
+  protected def _exceptions(): Set[ExceptionCodeGen] = Set.empty
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala:45]: We could make this abstract to clarify the difference between it and the non-underscore variant, but that would on the other hand mean duplicated `Set.empty` implementations in subclasses... I'm torn on this one.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/ProjectionInstruction.scala [hunk 9]
```diff
@@ -75,16 +78,24 @@ case class ProjectRelProperty(token: Option[Int], propName: String, relIdVar: St
   override def fields() = if (token.isEmpty) s"private int $propKeyVar = -1;" else ""
 
   def projectedVariable = JavaSymbol(s"ro.relationshipGetProperty( $relIdVar, $propKeyVar ).value( null )", OBJECT)
+
+  override protected def _exceptions(): Set[ExceptionCodeGen] = Set(KernelExceptionCodeGen)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/ProjectionInstruction.scala]: This and line 61 should be without return type annotation (IMHO).

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/ProjectionInstruction.scala [hunk 10]
```diff
@@ -150,6 +163,8 @@ case class ProjectSubtraction(lhs: ProjectionInstruction, rhs: ProjectionInstruc
 
   def fields() = ""
 
+  override def children = Seq(lhs, rhs)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/ProjectionInstruction.scala:166]: Ah, you found that bug too :)

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala [hunk 11]
```diff
@@ -66,26 +68,63 @@ abstract class CompiledExecutionResult(completion: CompletionListener, statement
...
     successful = true
   }
 
   override def close(): Unit = {
-    if (innerIterator != null) {
-      innerIterator.close( )
-    }
-    statement.close()
-    completion.complete(success=successful)
+    taskCloser.close(success = successful)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala:118]: I was told that in Scala convention, `=` is not used for functions which does not return any value (ie, java `void`). Concerns this and line 113. Also, `Unit` is superfluous.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala [hunk 12]
```diff
@@ -66,26 +68,63 @@ abstract class CompiledExecutionResult(completion: CompletionListener, statement
...
   }
 
   override def planDescriptionRequested: Boolean =  executionMode == ExplainMode || executionMode == ProfileMode
 
-  override def executionType: QueryExecutionType =  if (executionMode == ProfileMode) profiled(queryType) else query(queryType)
+  override def executionType = if ( executionMode == ProfileMode ) {
+    profiled( queryType )
+  } else {
+    query( queryType )
+  }
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala:127]: Some Java style parenthesis padding. Perhaps we want this?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala [hunk 13]
```diff
@@ -107,17 +147,68 @@ abstract class CompiledExecutionResult(completion: CompletionListener, statement
...
+
+  def text(a: Any): String = a match {
+    case x: Node            => x.toString + props(x)
+    case x: Relationship    => ":" + x.getType.name() + "[" + x.getId + "]" + props(x)
+    case x if x.isInstanceOf[Map[_, _]] => makeString(x.asInstanceOf[Map[String, Any]])
+    case x if x.isInstanceOf[java.util.Map[_, _]] => makeString(x.asInstanceOf[java.util.Map[String, Any]].asScala)
+    case IsCollection(coll) => coll.map(elem => text(elem)).mkString("[", ",", "]")
+    case x: String          => "\"" + x + "\""
+    case v: KeyToken        => v.name
+    case Some(x)            => x.toString
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala:189]: We match on `Some(x)`, but not on `None`. Is this correct?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/ExecutionPlanBuilder.scala [hunk 14]
```diff
@@ -82,10 +87,21 @@ class ExecutionPlanBuilder(graph: GraphDatabaseService, statsDivergenceThreshold
       def isStale(lastTxId: () => Long, statistics: GraphStatistics) = fingerprint.isStale(lastTxId, statistics)
 
       def run(queryContext: QueryContext, kernelStatement: KernelStatement,
-                       planType: ExecutionMode, params: Map[String, Any]): InternalExecutionResult =
-        compiledPlan.executionResultBuilder(kernelStatement, graph, planType, params, new CompletionListener {
-          override def complete(success: Boolean) = queryContext.close(success)
-        })
+              planType: ExecutionMode, params: Map[String, Any]): InternalExecutionResult = {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/ExecutionPlanBuilder.scala]: We should probably rename `planType` to something with `mode`. `executionMode`, `mode`, `queryMode`?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/formatOutput.scala [hunk 15]
```diff
@@ -80,21 +44,21 @@ object formatOutput extends ((ReadOperations, PrintWriter, List[String], Seq[Map
 
     def repeat(x: String, size: Int): String = (1 to size).map((i) => x).mkString
 
-    def createString(columnSizes: Map[String, Int], m: Map[String, Any]): String = {
+    def createString(columnSizes: Map[String, Int], m: Map[String, String]): String = {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/formatOutput.scala]: Return type annotation.

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/CodeGeneratorTest.scala [hunk 16]
```diff
@@ -422,23 +420,69 @@ class CodeGeneratorTest extends CypherFunSuite with LogicalPlanningTestSupport {
...
+        compiled.accept(visitor)
+        fail("should have thrown error")
+      }
+      catch {
+        case e: Throwable => e should equal(exception)
+      }
+
+  }
+
+  test("throw error when parameter is missing") {
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/CodeGeneratorTest.scala:460]: Given, when, then.

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/CodeGeneratorTest.scala [hunk 17]
```diff
@@ -422,23 +420,69 @@ class CodeGeneratorTest extends CypherFunSuite with LogicalPlanningTestSupport {
...
+    val plan = ProduceResult(List.empty, List.empty, List("a"), Projection(SingleRow()(solved), Map("a" -> SignedDecimalIntegerLiteral("1")(null)))(solved))
+
+    // when
+    val closer = mock[TaskCloser]
+    val compiled = compile( plan, taskCloser = closer )
+
+    // then
+    val visitor = mock[ResultVisitor[RuntimeException]]
+    val exception = new scala.RuntimeException()
+    when(visitor.visit(any[ResultRow])).thenThrow(exception)
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/CodeGeneratorTest.scala:449]: I don't understand this test. Why are we testing that the mocking framework throws the correct exception for us? Shouldn't we send bad input in and let the function throw by itself?

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/CodeGeneratorTest.scala [hunk 18]
```diff
@@ -422,23 +420,69 @@ class CodeGeneratorTest extends CypherFunSuite with LogicalPlanningTestSupport {
...
+    val name = """{"a":
+              |1
+              |}
+            """.stripMargin
+    val plan = ProduceResult(List.empty, List.empty, List(name), Projection(SingleRow()(solved),
+      Map(name -> SignedDecimalIntegerLiteral("1")(pos)))(solved))
+
+    val compiled = compile(plan, Map("FOO" -> Long.box(3L), "BAR" -> Long.box(1L)))
+
+    //then
```
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/CodeGeneratorTest.scala:483]: Given, when.

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/CodeGeneratorTest.scala (hunk 17 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/CodeGeneratorTest.scala:449]: we are testing that if a user sends in a `ResultsVisitor<MyAwesomeException>` and that visitor throws we should propagate that exception to the user and not hide it. The outer (java) signature is something like

```
<E extends Exception> void accept( ResultVisitor<E> visitor )
            throws E;
```

thus we must propagate `E`

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala (hunk 6 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala:40]: Having a final method an a trait seems a bit weird. However I do agree that this needs to be cleaned up but let's do that as separate PR.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala (hunk 7 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala:68]: as above let's clean this up in a later PR, if we want to change this we should change all the methods in this trait

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala (hunk 8 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala:45]: I agree but let's not do this now.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala (hunk 13 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala:189]: None will be matched on 192, `case x => x.toString`

File: community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/CodeGeneratorTest.scala (hunk 17 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/test/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/CodeGeneratorTest.scala:449]: I see. Thanks!

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala (hunk 6 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/Instruction.scala:40]: Fair enough.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala (hunk 13 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala:189]: Great!

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala (hunk 11 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala:118]: Not according to http://docs.scala-lang.org/style/declarations.html

**Comment by pontusmelke** [comment]
Failures are not related to this PR

**Comment by Mats-SX** [comment]
Does that mean that we can merge it then?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala (hunk 11 above)
Comment (Mats-SX) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/executionplan/CompiledExecutionResult.scala:118]: Oh. Comments on this, @systay?

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/codegen/LogicalPlanConverter.scala (hunk 5 above)
Comment (pontusmelke) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/codegen/LogicalPlanConverter.scala:185]: it is needed, called recursively
